py3status/modules/playerctl.py

```diff
@@ -0,0 +1,272 @@
+    }
+
+    class Meta:
+        update_config = {
+            "update_placeholder_format": [
+                {
+                    "placeholder_formats": {"title": ":escape"},
+                    "format_strings": ["format_player"],
+                }
+            ]
+        }
+
+    def post_config_hook(self):
+        self.color_paused = self.py3.COLOR_PAUSED or self.py3.COLOR_DEGRADED
+        self.color_playing = self.py3.COLOR_PLAYING or self.py3.COLOR_GOOD
+        self.color_stopped = self.py3.COLOR_STOPPED or self.py3.COLOR_BAD
+        self.thresholds_init = self.py3.get_color_names_list(self.format_player)
+
+        # Initialize the manager + event listeners
+        self.manager = Playerctl.PlayerManager()
+        self.manager.connect("name-appeared", self._on_name_appeared)
+        self.manager.connect("name-vanished", self._on_name_vanished)
+
+        # Initialize all currently active players
+        for player_name in self.manager.props.player_names:
+            self._init_player(player_name)
+
+        self._start_loop()
+
+    def _player_should_be_tracked(self, player_name):
+        for _filter in self.players:
+            if fnmatch(player_name.name, _filter):
+                return True
+
+        return len(self.players) == 0
+
+    def _init_player(self, player_name):
+        if not self._player_should_be_tracked(player_name):
+            return
+
+        player = Playerctl.Player.new_from_name(player_name)
+
+        # Initialize player event listeners
+        player.connect("playback-status", self._status_changed, self.manager)
+        player.connect("metadata", self._on_metadata, self.manager)
+
+        self.manager.manage_player(player)
+
+    def _start_loop(self):
+        loop = GLib.MainLoop()
+
+        # The loop is blocking so it needs to be run a separate thread
+        thread = Thread(target=loop.run)
+        thread.daemon = True
+        thread.start()
+
+    def _on_name_appeared(self, manager, player_name):
+        self._init_player(player_name)
+
+    def _on_name_vanished(self, manager, name):
+        # Add a very small delay to give Playerctl time to remove the player
+        time.sleep(0.01)
+        self.py3.update()
+
+    def _on_metadata(self, player, metadata, manager):
+        self.py3.update()
+
+    def _status_changed(self, player, status, manager):
+        self.py3.update()
+
+    @staticmethod
+    def _microseconds_to_time(microseconds):
+        seconds = microseconds // 1_000_000
+        m, s = divmod(seconds, 60)
+        h, m = divmod(m, 60)
+        time = f"{h}:{m:02d}:{s:02d}"
+        return time.lstrip("0").lstrip(":")
+
+    def _get_player_position(self, player):
+        try:
+            # Playerctl doesn't support getting positions for all players
+            microseconds = player.get_position()
+            position = self._microseconds_to_time(microseconds)
+        except GLib.GError:
+            position = None
+
+        return position
```

Comment:
(Comment) I thought for a brief second to make even smarter `cache_timeout` here... via using `any([pos1, pos2, pos3])`... but I think that's an overkill because 9 out of 10 decent players will have positions.

Talking about getting at least two players with no position. Probably won't last long anyway.

Reply:
What can be done is checking if any of the players are playing and have a valid position. This would be the only case I see that `cache_timeout` shouldn't be `CACHE_FOREVER`.

Reply:
It's your decision. Worth it or not. Idk.